initial rock integration

conrad-vanl opened the pull request:
Fixes #5 

## Creator

- [ ] Build relevant tests if any apply
- [ ] Ensure there are no new warnings (in app AND in storybook)
- [ ] Upload an animated GIF showcasing the solution (if it applies)
- [ ] Set a relevant reviewer

## Reviewer

- [ ] Review CI build results
- [ ] Review function and form on iOS
- [ ] Review function and form on Android
- [ ] Review new stories if any apply
- [ ] Review test coverage: Are all new or changed states tested?
- [ ] Review new test logic

conrad-vanl commented:
@IsaacHardy if I try to run `yarn lint` from this branch locally, it just seems to hang endlessly. Did you test the linter locally before? Trying to figure out what could be going on...

IsaacHardy commented:
@conrad-vanl Looking at this now.

IsaacHardy commented:
@conrad-vanl Just saw your comment above. Yes, I tested locally multiple times.

IsaacHardy commented on api/src/data/content-channels/index.js:
```diff
@@ -0,0 +1,33 @@
+import {
```
Are we waiting on a schema to make tests?

conrad-vanl replied:
....yes? 😇

IsaacHardy commented on api/src/data/index.js:
```diff
@@ -0,0 +1,28 @@
+    node(id: ID!): Node
+    contentChannels: [ContentChannel]
+  }
+  ${values(data).map(datum => datum.schema)}
```
I like this.

conrad-vanl commented on api/src/connectors/rest.js:
```diff
@@ -0,0 +1,106 @@
+    if (Array.isArray(data)) return data.map(this.normalize);
+    if (typeof data !== 'object') return data;
+    const normalizedValues = mapValues(data, this.normalize);
+    return mapKeys(normalizedValues, (v, key) => camelCase(key));
```
100% agree. I don't think we will be directly exposing the Rock API schema (or any API, for that matter), and these parameters only represent the parameter names coming from the Rock API. We normalize them only because each API we connect with will likely utilize a different capitalization/formatting scheme for parameters, and this makes it consistent within our codebase (dasherized, camelCase, PascalCase, underscored, etc all get converted to camelCase). But our resolvers and graphql schema will be where our naming matters.

IsaacHardy replied:
I like all you've said here and agree, but I just wanted the `v` to be changed to `value`. 😜

conrad-vanl replied:
OOOOOOOH that's what you were talking about, lolz. WORDZ. Yeah, good catch. I'll change.

conrad-vanl commented on api/src/connectors/rest.test.js:
```diff
@@ -0,0 +1,189 @@
+import RestConnector, {
```
🤩

conrad-vanl commented:
Alright I addressed your comments @IsaacHardy. 

> Also, what is the plan with the content-channels? Just create the initial setup for now? Build it out as we work with the Rock team to develop a compatible schema?

Correct. The primary purpose of this PR is to port over `node` from Heighliner and setup our `connector`/`model`/`resolver` paradigm. But I do not believe anything in content channels will be kept - which is why I just did `id`, `title` and `description` for now (and there's a chance `id` is the only thing kept as-is as we build out `content-channels`).

There will be a separate PR / issue discussing and building the exact schema for content channels.

> Lastly, I am only read a little bit about Relay's global object identification approach which I believe we are mimicking here in Apollo and would love a better explanation of its purpose and use-fullness.

Correct. Furthermore, I actually copied over `node` almost verbatim from `Heighliner`. It was pretty well done IMO. As far as this pattern pertains to us, the positive to globally unique IDs is it turns the IDs into a transparent data structure to the client. We can embed any information we want in the ID and know we'll get that information back, and the client doesn't know about it. This has numerous benefits - like [caching](https://graphql.org/learn/caching/). In the future, I could see the need to embed not only the object type, but also the data source it came from (like Rock) to help control against customizable data connectors/sources. For now, it's a good practice to follow, it helps us with caching, and allows us to expose a single root-level `node` query that can get any object the graphql api exposes by ID.

IsaacHardy commented on api/src/connectors/rock.js:
```diff
@@ -0,0 +1,18 @@
+import RestConnector from './rest';
+
+const ROCK_API = process.env.ROCK_API || 'https://apollosrock.newspring.cc/api';
+const ROCK_TOKEN = process.env.ROCK_TOKEN || 't1LM3cVMJG2sc2jKa6nVLld8'; // TODO: Don't commit this :)
```
Yea. Let's invalidate it. I missed it, too. I have begin looking into how to setup `.env` files yet, but that is definitely on my plate after unit tests and PR/code review standards.
